### ISSUE TITLE
Addition: Reasonml

### ISSUE THREAD

**Issue by Haroenv** [github/linguist#3949]
There seems to be a file with it already in the vendors folder, but it's not picked up on GitHub.com. Which files need to be edited for it to be possible?

**Comment by lildude** [comment]
Reason should be detected. If you're not seeing it being detected, there's likely another explanation.

Can you provide a link to a repository where you are seeing that it isn't being detected.

**Comment by Haroenv** [comment]
* in https://github.com/reasonml-community/reason-react-hacker-news/ it's detected as OCaml or C++
* https://github.com/BuckleScript/bucklescript/blob/master/jscomp/build_tests/jsxv/src/demo.re doesn't seem to be detected
* https://github.com/reasonml-community/bs-react-native has detected as Reason in the search, but not in the overview

I've never seen Reason in the language overview

**Comment by Haroenv** [comment]
* https://github.com/reasonml-community/bs-director
* https://github.com/reasonml-community/reductive

and a bunch of other projects on https://github.com/reasonml-community as well

**Comment by pchaigno** [comment]
Reason is [part of the OCaml group](https://github.com/github/linguist/blob/d4c2d83af9f576aacdd97d58796d2a22fa90def4/lib/linguist/languages.yml#L3813) so Reason files are counted as OCaml files (only the syntax highlighting differs).

**Comment by Haroenv** [comment]
Interesting, I knew it's a different syntax for the same language, but I'm wondering if it wouldn't be better if it was split out in its own language colour

**Comment by mlms13** [comment]
I agree that it would be nice for Reason to be shown separately from OCaml, but independent of that, I can also confirm that some Reason files are getting misidentified as C++.

**Comment by lpil** [comment]
Given ReasonML is just an alternative parser for the OCaml compiler I don't think it should be a different language.

**Comment by Haroenv** [comment]
* https://github.com/reasonml-community/reason-react-hacker-news/search?l=c%2B%2B detected as c++ still

Not sure if you want to fix this still

**Comment by lildude** [comment]
Thanks for the nudge @Haroenv. Looks like this is caused by a lack of a heuristic for the `.re` extension. It is associated with both C++ and Reason but without an heuristic so is falling through to the classifier.

I have no experience with either language, but if you've got the bandwidth and some experience, we'd appreciate a PR adding the necessary heuristic to distinguish the two languages.

**Comment by Haroenv** [comment]
Sorry, not my area of expertise :)

**Comment by Haroenv** [comment]
I'll save stale bot and close it myself since I don't mind enough to dig into how linguist works to fix this. Someone else of course can open a new issue for this c++ / reason misreading problem

**Comment by dawee** [comment]
I really think ReasonML projects should have their own identity. Reason is a growing community and yes it's an extension to OCaml, but OCaml itself is one way to parse OCaml, it's meta but it doesn't mean it's not something new.

Here are the reason (cough cough), I think, it should be the language detected:

1) I've been a C++ developer for several years. I saw cxx, cpp, hpp but re, never. Before writing here I checked on cppreference and I found no reference to ".re" files for c++.

2) We are right now in this situation that seems like a bug to me:

![trending-reason](https://user-images.githubusercontent.com/1039076/54860972-6022bb00-4cf8-11e9-8fbf-7e26fb14b15e.png)

I can choose Reason in the languages list but I have no way to have anything in this feed because all the reason projects are saved as OCaml/C++

3) Typescript is recognized as a language. It's a typing subset to javascript. I don't see why it would be considered as a new language if Reason can't.

Trending repositories is a tool I use every day to see what communities are bringing. Reason is a fairly new community but a lot of new stuff comes up and I can't follow it like I do with javascript, rust, python and all the other communities I like.
